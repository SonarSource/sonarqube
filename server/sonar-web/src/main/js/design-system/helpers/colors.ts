/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { CSSColor } from '../types/theme';

/* eslint-disable no-bitwise, no-mixed-operators */

export function stringToColor(str: string) {
  let hash = 0;

  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }

  let color = '#';

  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += value.toString(16).padStart(2, '0');
  }

  return color;
}

export function isDarkColor(color: string) {
  color = color.substring(1);

  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }

  const rgb = parseInt(color.substring(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;

  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}

export function getRGBAString([r, g, b]: Array<number | string>, a?: number | string) {
  return (a !== undefined ? `rgba(${r},${g},${b},${a})` : `rgb(${r},${g},${b})`) as CSSColor;
}
