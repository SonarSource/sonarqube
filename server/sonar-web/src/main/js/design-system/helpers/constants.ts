/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { theme } from 'twin.macro';

export const DEFAULT_LOCALE = 'en';
export const IS_SSR = typeof window === 'undefined';
export const REACT_DOM_CONTAINER = '#content';

export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const THROTTLE_SCROLL_DELAY = 10;
export const THROTTLE_KEYPRESS_DELAY = 100;

export const DEBOUNCE_DELAY = 250;

export const DEBOUNCE_LONG_DELAY = 1000;

export const DEBOUNCE_SUCCESS_DELAY = 1000;

export const INTERACTIVE_TOOLTIP_DELAY = 0.5;

export const LEAK_PERIOD = 'sonar.leak.period';

export const LEAK_PERIOD_TYPE = 'sonar.leak.period.type';

export const INPUT_SIZES = {
  small: theme('width.input-small'),
  medium: theme('width.input-medium'),
  large: theme('width.input-large'),
  full: theme('width.full'),
  auto: theme('width.auto'),
};

export const LAYOUT_VIEWPORT_MIN_WIDTH = 1280;
export const LAYOUT_VIEWPORT_MAX_WIDTH = 1280;
export const LAYOUT_VIEWPORT_MAX_WIDTH_LARGE = 1680;
export const LAYOUT_MAIN_CONTENT_GUTTER = 60;
export const LAYOUT_SIDEBAR_WIDTH = 240;
export const LAYOUT_SIDEBAR_COLLAPSED_WIDTH = 60;
export const LAYOUT_SIDEBAR_BREAKPOINT = 1320;
export const LAYOUT_BANNER_HEIGHT = 44;
export const LAYOUT_BRANDING_ICON_WIDTH = 198;
export const LAYOUT_FILTERBAR_HEADER = 56;
export const LAYOUT_GLOBAL_NAV_HEIGHT = 52;
export const LAYOUT_PROJECT_NAV_HEIGHT = 108;
export const LAYOUT_LOGO_MARGIN_RIGHT = 45;
export const LAYOUT_LOGO_MAX_HEIGHT = 40;
export const LAYOUT_LOGO_MAX_WIDTH = 150;
export const LAYOUT_FOOTER_HEIGHT = 69;
export const LAYOUT_NOTIFICATIONSBAR_WIDTH = 350;

export const CORE_CONCEPTS_WIDTH = 350;

export const DARK_THEME_ID = 'dark-theme';

export const OPACITY_20_PERCENT = 0.2;

export const OPACITY_75_PERCENT = 0.75;

export const GLOBAL_POPUP_Z_INDEX = 5000;

export const TOAST_AUTOCLOSE_DELAY = 5000;
