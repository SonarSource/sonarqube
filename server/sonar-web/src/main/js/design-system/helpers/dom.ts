/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

/**
 * This function find the point on the target element using the rect coordinat.
 * This point will serve as an anchor for rect to be attached
 *
 * This function assume that rect side will overlap the target facing side. If not
 * the case the point would be outside the target rect. For now we don't need to
 * handle this situation.
 * @param rect
 * @param targetRect
 */
export function findAnchor(rect: DOMRect, targetRect: DOMRect, offset: number) {
  const offestTop = rect.top < targetRect.top ? targetRect.top - rect.top : 0;
  const offestLeft = rect.left < targetRect.left ? targetRect.left - rect.left : 0;

  if (targetRect.right < rect.left) {
    const left = targetRect.right - rect.left - offset / 2;
    const top =
      (Math.min(targetRect.bottom, rect.bottom) - Math.max(targetRect.top, rect.top)) / 2 -
      offset / 2 +
      offestTop;
    const rotate = '0deg';
    const width = -left - offset;

    return { left, top, rotate, width };
  } else if (targetRect.left > rect.right) {
    const left = rect.width + targetRect.left - rect.right + offset / 2;
    const top =
      (Math.min(targetRect.bottom, rect.bottom) - Math.max(targetRect.top, rect.top)) / 2 -
      offset / 2 +
      offestTop;
    const rotate = '180deg';
    const width = left - rect.width;
    return { left, top, rotate, width };
  } else if (targetRect.bottom < rect.top) {
    const left =
      (Math.min(targetRect.right, rect.right) - Math.max(targetRect.left, rect.left)) / 2 -
      offset / 2 +
      offestLeft;
    const top = targetRect.bottom - rect.top - offset / 2;
    const rotate = '90deg';
    const width = -top - offset;
    return { left, top, rotate, width };
  } else if (targetRect.top > rect.bottom) {
    const left =
      (Math.min(targetRect.right, rect.right) - Math.max(targetRect.left, rect.left)) / 2 -
      offset / 2 +
      offestLeft;
    const top = targetRect.top - rect.top - offset / 2;
    const rotate = '-90deg';
    const width = top - rect.height;
    return { left, top, rotate, width };
  }

  // When rectagle overlap
  return { left: 0, top: 0, rotate: '0deg', width: 0 };
}
