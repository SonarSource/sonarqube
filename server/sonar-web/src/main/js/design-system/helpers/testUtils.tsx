/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { Options as UserEventsOptions } from '@testing-library/user-event/dist/types/options';
import { InitialEntry } from 'history';
import { identity, kebabCase } from 'lodash';
import React, { PropsWithChildren, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { IntlProvider, ReactIntlErrorCode } from 'react-intl';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

type RenderResultWithUser = RenderResult & { user: UserEvent };

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
): RenderResultWithUser {
  return { ...rtlRender(ui, options), user: userEvent.setup(userEventOptions) };
}

type RenderContextOptions = Omit<RenderOptions, 'wrapper'> & {
  initialEntries?: InitialEntry[];
  userEventOptions?: UserEventsOptions;
};

export function renderWithContext(
  ui: React.ReactElement,
  { userEventOptions, ...options }: RenderContextOptions = {},
) {
  return render(ui, { ...options, wrapper: getContextWrapper() }, userEventOptions);
}

interface RenderRouterOptions {
  additionalRoutes?: ReactNode;
}

export function renderWithRouter(
  ui: React.ReactElement,
  options: RenderContextOptions & RenderRouterOptions = {},
) {
  const { additionalRoutes, userEventOptions, ...renderOptions } = options;

  function RouterWrapper({ children }: React.PropsWithChildren<object>) {
    return (
      <HelmetProvider>
        <IntlWrapper>
          <MemoryRouter>
            <Routes>
              <Route element={children} path="/" />
              {additionalRoutes}
            </Routes>
          </MemoryRouter>
        </IntlWrapper>
      </HelmetProvider>
    );
  }

  return render(ui, { ...renderOptions, wrapper: RouterWrapper }, userEventOptions);
}

function getContextWrapper() {
  return function ContextWrapper({ children }: React.PropsWithChildren<object>) {
    return (
      <HelmetProvider>
        <IntlWrapper>{children}</IntlWrapper>
      </HelmetProvider>
    );
  };
}

export function mockComponent(name: string, transformProps: (props: any) => any = identity) {
  function MockedComponent({ ...props }: PropsWithChildren<any>) {
    return React.createElement('mocked-' + kebabCase(name), transformProps(props));
  }

  MockedComponent.displayName = `mocked(${name})`;
  return MockedComponent;
}

export const debounceTimer = jest
  .fn()
  .mockImplementation((callback: (...args: unknown[]) => void, timeout: number) => {
    let timeoutId: number;

    const debounced = jest.fn((...args: unknown[]) => {
      window.clearTimeout(timeoutId);

      timeoutId = window.setTimeout(() => {
        callback(...args);
      }, timeout);
    });

    (debounced as typeof debounced & { cancel: () => void }).cancel = jest.fn(() => {
      window.clearTimeout(timeoutId);
    });

    return debounced;
  });

export function IntlWrapper({
  children,
  messages = {},
}: {
  children: ReactNode;
  messages?: Record<string, string>;
}) {
  return (
    <IntlProvider
      defaultLocale="en"
      locale="en"
      messages={messages}
      onError={(e) => {
        // ignore missing translations, there are none!
        if (
          e.code !== ReactIntlErrorCode.MISSING_TRANSLATION &&
          e.code !== ReactIntlErrorCode.UNSUPPORTED_FORMATTER
        ) {
          // eslint-disable-next-line no-console
          console.error(e);
        }
      }}
    >
      {children}
    </IntlProvider>
  );
}
