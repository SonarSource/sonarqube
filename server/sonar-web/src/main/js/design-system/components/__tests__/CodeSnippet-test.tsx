/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { screen } from '@testing-library/react';
import { HelmetProvider } from 'react-helmet-async';
import { renderWithContext } from '../../helpers/testUtils';
import { FCProps } from '../../types/misc';
import { CodeSnippet } from '../CodeSnippet';

it('should show full size when multiline with no editing', () => {
  const { container } = setupWithProps();
  const copyButton = screen.getByRole('button', { name: 'Copy' });
  expect(copyButton).toHaveStyle('top: 1.5rem');
  expect(container).toMatchSnapshot();
});

it('should show reduced size when single line with no editing', () => {
  const { container } = setupWithProps({ isOneLine: true, snippet: 'foobar' });
  const copyButton = screen.getByRole('button', { name: 'Copy' });
  expect(copyButton).toHaveStyle('top: 1rem');
  expect(container).toMatchSnapshot();
});

it('should highlight code content correctly', () => {
  const { container } = setupWithProps({ snippet: '<prop>foobar<prop>' });
  expect(container).toMatchSnapshot();
});

function setupWithProps(props: Partial<FCProps<typeof CodeSnippet>> = {}) {
  return renderWithContext(
    <HelmetProvider>
      <CodeSnippet snippet={'foo\nbar'} {...props} />
    </HelmetProvider>,
  );
}
