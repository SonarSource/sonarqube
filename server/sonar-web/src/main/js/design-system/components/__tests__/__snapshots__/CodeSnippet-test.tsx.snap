// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should highlight code content correctly 1`] = `
.emotion-0 {
  background-color: rgb(252,252,253);
  border: 1px solid rgb(225,230,243);
  border-radius: 0.5rem;
  position: relative;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.emotion-0.code-snippet-simple-oneline {
  margin-top: 0;
  margin-bottom: 0;
  border-radius: 0.25rem;
}

.emotion-3 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: var(--echoes-dimension-space-0) var(--button-padding);
  height: var(--button-height);
  min-height: var(--button-height);
  overflow: hidden;
  font: var(--echoes-typography-others-label);
  color: var(--button-color);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--button-background);
  border: var(--button-border);
  border-radius: var(--echoes-border-radius-400);
  outline: none;
  cursor: pointer;
  --button-color: var(--echoes-color-text-default);
  --button-border: var(--echoes-color-border-bold) solid var(--echoes-border-width-default);
  --button-background: var(--echoes-color-background-default);
  --button-background-hover: var(--echoes-color-background-default-hover);
  --button-background-active: var(--echoes-color-background-default-active);
  --button-background-focus: var(--echoes-color-background-default);
  --button-background-disabled: var(--echoes-color-background-disabled);
  --button-padding: var(--echoes-dimension-space-150);
  --button-height: var(--echoes-sizes-buttons-large);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font: var(--echoes-typography-text-default-regular);
  right: 1.5rem;
  top: 1.5rem;
  position: absolute;
}

.emotion-3:focus,
.emotion-3:focus-visible {
  background-color: var(--button-background-focus);
  outline: var(--echoes-color-focus-default) solid var(--echoes-focus-border-width-default);
  outline-offset: var(--echoes-focus-border-offset-default);
}

.emotion-3:hover {
  background-color: var(--button-background-hover);
}

.emotion-3:active {
  background-color: var(--button-background-active);
}

.emotion-3:disabled,
.emotion-3:disabled:has(:hover, :active, :focus, :focus-visible) {
  color: var(--echoes-color-text-disabled);
  background-color: var(--button-background-disabled);
  border: none;
  cursor: not-allowed;
  pointer-events: none;
}

.code-snippet-highlighted-oneline .emotion-3 {
  bottom: 0.5rem;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--echoes-dimension-space-75);
  overflow: hidden;
}

.emotion-7 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
  font-family: 'Material Symbols Rounded';
}

.emotion-9 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-11 code {
  font: var(--echoes-typography-code-default);
  background: rgb(252,252,253);
  color: rgb(51,53,60);
}

.emotion-11 code.hljs {
  padding: unset;
}

.emotion-11 .hljs-meta,
.emotion-11 .hljs-variable {
  color: rgb(51,53,60);
}

.emotion-11 .hljs-doctag,
.emotion-11 .hljs-title,
.emotion-11 .hljs-title.class_,
.emotion-11 .hljs-title.function_ {
  color: rgb(34,84,192);
}

.emotion-11 .hljs-comment {
  font: var(--echoes-typography-code-comment);
  color: rgb(109,111,119);
}

.emotion-11 .hljs-keyword,
.emotion-11 .hljs-tag,
.emotion-11 .hljs-type {
  color: rgb(152,29,150);
}

.emotion-11 .hljs-literal,
.emotion-11 .hljs-number {
  color: rgb(126,83,5);
}

.emotion-11 .hljs-string {
  color: rgb(32,105,31);
}

.emotion-11 .hljs-meta .hljs-keyword {
  color: rgb(47,103,48);
}

.emotion-11 .sonar-underline {
  -webkit-text-decoration: underline rgb(253,162,155);
  text-decoration: underline rgb(253,162,155);
  -webkit-text-decoration: underline rgb(253,162,155) wavy;
  text-decoration: underline rgb(253,162,155) wavy;
  text-decoration-thickness: 2px;
  text-decoration-skip-ink: none;
}

.emotion-11.code-wrap {
  white-space: pre-wrap;
  word-break: break-all;
}

.emotion-11.code-wrap.wrap-words {
  word-break: normal;
  overflow-wrap: break-word;
}

.emotion-11 mark {
  font-weight: 400;
  padding: 0.25rem;
  border-radius: 0.25rem;
  background-color: rgb(197,205,223);
  color: rgb(217,45,32);
}

<div>
  <div
    class="sw-code fs-mask emotion-0 emotion-1"
  >
    <button
      class="sw-select-none emotion-2 emotion-3 emotion-4"
      data-state="closed"
      type="button"
    >
      <span
        class="emotion-5 emotion-6"
      >
        <span
          aria-hidden="true"
          class="emotion-7 emotion-8"
        >
          ÓÖç
        </span>
        <span
          class="emotion-9 emotion-10"
        >
          Copy
        </span>
      </span>
    </button>
    <span
      class="hljs sw-overflow-auto sw-pr-24 sw-flex emotion-11 emotion-12"
    >
      <pre>
        &lt;prop&gt;foobar&lt;prop&gt;
      </pre>
    </span>
  </div>
</div>
`;

exports[`should show full size when multiline with no editing 1`] = `
.emotion-0 {
  background-color: rgb(252,252,253);
  border: 1px solid rgb(225,230,243);
  border-radius: 0.5rem;
  position: relative;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.emotion-0.code-snippet-simple-oneline {
  margin-top: 0;
  margin-bottom: 0;
  border-radius: 0.25rem;
}

.emotion-3 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: var(--echoes-dimension-space-0) var(--button-padding);
  height: var(--button-height);
  min-height: var(--button-height);
  overflow: hidden;
  font: var(--echoes-typography-others-label);
  color: var(--button-color);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--button-background);
  border: var(--button-border);
  border-radius: var(--echoes-border-radius-400);
  outline: none;
  cursor: pointer;
  --button-color: var(--echoes-color-text-default);
  --button-border: var(--echoes-color-border-bold) solid var(--echoes-border-width-default);
  --button-background: var(--echoes-color-background-default);
  --button-background-hover: var(--echoes-color-background-default-hover);
  --button-background-active: var(--echoes-color-background-default-active);
  --button-background-focus: var(--echoes-color-background-default);
  --button-background-disabled: var(--echoes-color-background-disabled);
  --button-padding: var(--echoes-dimension-space-150);
  --button-height: var(--echoes-sizes-buttons-large);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font: var(--echoes-typography-text-default-regular);
  right: 1.5rem;
  top: 1.5rem;
  position: absolute;
}

.emotion-3:focus,
.emotion-3:focus-visible {
  background-color: var(--button-background-focus);
  outline: var(--echoes-color-focus-default) solid var(--echoes-focus-border-width-default);
  outline-offset: var(--echoes-focus-border-offset-default);
}

.emotion-3:hover {
  background-color: var(--button-background-hover);
}

.emotion-3:active {
  background-color: var(--button-background-active);
}

.emotion-3:disabled,
.emotion-3:disabled:has(:hover, :active, :focus, :focus-visible) {
  color: var(--echoes-color-text-disabled);
  background-color: var(--button-background-disabled);
  border: none;
  cursor: not-allowed;
  pointer-events: none;
}

.code-snippet-highlighted-oneline .emotion-3 {
  bottom: 0.5rem;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--echoes-dimension-space-75);
  overflow: hidden;
}

.emotion-7 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
  font-family: 'Material Symbols Rounded';
}

.emotion-9 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-11 code {
  font: var(--echoes-typography-code-default);
  background: rgb(252,252,253);
  color: rgb(51,53,60);
}

.emotion-11 code.hljs {
  padding: unset;
}

.emotion-11 .hljs-meta,
.emotion-11 .hljs-variable {
  color: rgb(51,53,60);
}

.emotion-11 .hljs-doctag,
.emotion-11 .hljs-title,
.emotion-11 .hljs-title.class_,
.emotion-11 .hljs-title.function_ {
  color: rgb(34,84,192);
}

.emotion-11 .hljs-comment {
  font: var(--echoes-typography-code-comment);
  color: rgb(109,111,119);
}

.emotion-11 .hljs-keyword,
.emotion-11 .hljs-tag,
.emotion-11 .hljs-type {
  color: rgb(152,29,150);
}

.emotion-11 .hljs-literal,
.emotion-11 .hljs-number {
  color: rgb(126,83,5);
}

.emotion-11 .hljs-string {
  color: rgb(32,105,31);
}

.emotion-11 .hljs-meta .hljs-keyword {
  color: rgb(47,103,48);
}

.emotion-11 .sonar-underline {
  -webkit-text-decoration: underline rgb(253,162,155);
  text-decoration: underline rgb(253,162,155);
  -webkit-text-decoration: underline rgb(253,162,155) wavy;
  text-decoration: underline rgb(253,162,155) wavy;
  text-decoration-thickness: 2px;
  text-decoration-skip-ink: none;
}

.emotion-11.code-wrap {
  white-space: pre-wrap;
  word-break: break-all;
}

.emotion-11.code-wrap.wrap-words {
  word-break: normal;
  overflow-wrap: break-word;
}

.emotion-11 mark {
  font-weight: 400;
  padding: 0.25rem;
  border-radius: 0.25rem;
  background-color: rgb(197,205,223);
  color: rgb(217,45,32);
}

<div>
  <div
    class="sw-code fs-mask emotion-0 emotion-1"
  >
    <button
      class="sw-select-none emotion-2 emotion-3 emotion-4"
      data-state="closed"
      type="button"
    >
      <span
        class="emotion-5 emotion-6"
      >
        <span
          aria-hidden="true"
          class="emotion-7 emotion-8"
        >
          ÓÖç
        </span>
        <span
          class="emotion-9 emotion-10"
        >
          Copy
        </span>
      </span>
    </button>
    <span
      class="hljs sw-overflow-auto sw-pr-24 sw-flex emotion-11 emotion-12"
    >
      <pre>
        foo
bar
      </pre>
    </span>
  </div>
</div>
`;

exports[`should show reduced size when single line with no editing 1`] = `
.emotion-0 {
  background-color: rgb(252,252,253);
  border: 1px solid rgb(225,230,243);
  border-radius: 0.5rem;
  position: relative;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.emotion-0.code-snippet-simple-oneline {
  margin-top: 0;
  margin-bottom: 0;
  border-radius: 0.25rem;
}

.emotion-3 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: var(--echoes-dimension-space-0) var(--button-padding);
  height: var(--button-height);
  min-height: var(--button-height);
  overflow: hidden;
  font: var(--echoes-typography-others-label);
  color: var(--button-color);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--button-background);
  border: var(--button-border);
  border-radius: var(--echoes-border-radius-400);
  outline: none;
  cursor: pointer;
  --button-color: var(--echoes-color-text-default);
  --button-border: var(--echoes-color-border-bold) solid var(--echoes-border-width-default);
  --button-background: var(--echoes-color-background-default);
  --button-background-hover: var(--echoes-color-background-default-hover);
  --button-background-active: var(--echoes-color-background-default-active);
  --button-background-focus: var(--echoes-color-background-default);
  --button-background-disabled: var(--echoes-color-background-disabled);
  --button-padding: var(--echoes-dimension-space-150);
  --button-height: var(--echoes-sizes-buttons-large);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font: var(--echoes-typography-text-default-regular);
  right: 1.5rem;
  top: 1.5rem;
  position: absolute;
  bottom: 1rem;
  top: 1rem;
}

.emotion-3:focus,
.emotion-3:focus-visible {
  background-color: var(--button-background-focus);
  outline: var(--echoes-color-focus-default) solid var(--echoes-focus-border-width-default);
  outline-offset: var(--echoes-focus-border-offset-default);
}

.emotion-3:hover {
  background-color: var(--button-background-hover);
}

.emotion-3:active {
  background-color: var(--button-background-active);
}

.emotion-3:disabled,
.emotion-3:disabled:has(:hover, :active, :focus, :focus-visible) {
  color: var(--echoes-color-text-disabled);
  background-color: var(--button-background-disabled);
  border: none;
  cursor: not-allowed;
  pointer-events: none;
}

.code-snippet-highlighted-oneline .emotion-3 {
  bottom: 0.5rem;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--echoes-dimension-space-75);
  overflow: hidden;
}

.emotion-7 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
  font-family: 'Material Symbols Rounded';
}

.emotion-9 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-11 code {
  font: var(--echoes-typography-code-default);
  background: rgb(252,252,253);
  color: rgb(51,53,60);
}

.emotion-11 code.hljs {
  padding: unset;
}

.emotion-11 .hljs-meta,
.emotion-11 .hljs-variable {
  color: rgb(51,53,60);
}

.emotion-11 .hljs-doctag,
.emotion-11 .hljs-title,
.emotion-11 .hljs-title.class_,
.emotion-11 .hljs-title.function_ {
  color: rgb(34,84,192);
}

.emotion-11 .hljs-comment {
  font: var(--echoes-typography-code-comment);
  color: rgb(109,111,119);
}

.emotion-11 .hljs-keyword,
.emotion-11 .hljs-tag,
.emotion-11 .hljs-type {
  color: rgb(152,29,150);
}

.emotion-11 .hljs-literal,
.emotion-11 .hljs-number {
  color: rgb(126,83,5);
}

.emotion-11 .hljs-string {
  color: rgb(32,105,31);
}

.emotion-11 .hljs-meta .hljs-keyword {
  color: rgb(47,103,48);
}

.emotion-11 .sonar-underline {
  -webkit-text-decoration: underline rgb(253,162,155);
  text-decoration: underline rgb(253,162,155);
  -webkit-text-decoration: underline rgb(253,162,155) wavy;
  text-decoration: underline rgb(253,162,155) wavy;
  text-decoration-thickness: 2px;
  text-decoration-skip-ink: none;
}

.emotion-11.code-wrap {
  white-space: pre-wrap;
  word-break: break-all;
}

.emotion-11.code-wrap.wrap-words {
  word-break: normal;
  overflow-wrap: break-word;
}

.emotion-11 mark {
  font-weight: 400;
  padding: 0.25rem;
  border-radius: 0.25rem;
  background-color: rgb(197,205,223);
  color: rgb(217,45,32);
}

<div>
  <div
    class="sw-code sw-py-6 code-snippet-highlighted-oneline fs-mask emotion-0 emotion-1"
  >
    <button
      class="sw-select-none emotion-2 emotion-3 emotion-4"
      data-state="closed"
      type="button"
    >
      <span
        class="emotion-5 emotion-6"
      >
        <span
          aria-hidden="true"
          class="emotion-7 emotion-8"
        >
          ÓÖç
        </span>
        <span
          class="emotion-9 emotion-10"
        >
          Copy
        </span>
      </span>
    </button>
    <span
      class="hljs sw-overflow-auto sw-pr-24 sw-flex emotion-11 emotion-12"
    >
      <code>
        foobar
      </code>
    </span>
  </div>
</div>
`;
